pkg/provisioner/ironic/ironic.go

```diff
@@ -1129,6 +1129,91 @@ func (p *ironicProvisioner) Adopt(force bool) (result provisioner.Result, err er
 	return
 }
 
+func (p *ironicProvisioner) startManualCleaning(ironicNode *nodes.Node) (result provisioner.Result, err error) {
+	var cleanSteps []nodes.CleanStep
+
+	if p.bmcAccess.RAIDInterface() != "" {
+		// Set raid configuration
+		err = setTargetRAIDCfg(p, ironicNode)
+		if err != nil {
+			p.log.Error(err, "failed in setting the target RAID configuration")
+			return
+		}
+		// Build raid clean steps
+		cleanSteps = append(cleanSteps, buildRAIDCleanSteps(p.host.Status.Provisioning.RAID)...)
+	} else if p.host.Status.Provisioning.RAID != nil {
+		return result, fmt.Errorf("the BMC(%v) of the current node does not support RAID", p.bmcAccess.Driver())
+	}
+
+	// Start manual clean
+	if len(cleanSteps) != 0 {
+		p.log.Info("remove existing configuration and set new configuration", "steps", cleanSteps)
+		opts := nodes.ProvisionStateOpts{Target: nodes.TargetClean, CleanSteps: cleanSteps}
+		var success bool
+		success, result, err = p.tryChangeNodeProvisionState(ironicNode, opts)
+		if !success {
+			err = fmt.Errorf("start manual clean failed")
+		}
+	}
+	return
+}
+
+// Prepare remove existing configuration and set new configuration
+func (p *ironicProvisioner) Prepare() (result provisioner.Result, err error) {
+	var ironicNode *nodes.Node
+
+	if ironicNode, err = p.findExistingHost(); err != nil {
+		return result, errors.Wrap(err, "could not find host to clean")
+	}
+	if ironicNode == nil {
+		return result, fmt.Errorf("no ironic node for host")
+	}
+
+	switch nodes.ProvisionState(ironicNode.ProvisionState) {
+	case nodes.Available, nodes.Active:
+		// Do Nothing (go to provisioning state)
+
+	case nodes.Manageable:
+		if p.host.Status.Cleaned {
+			// Manual clean finished
+			return
+		}
+		result, err = p.startManualCleaning(ironicNode)
+		if err == nil {
+			p.host.Status.Cleaned = true
+		}
+		return
+
+	case nodes.CleanFail:
+		if p.host.Status.Cleaned {
+			p.host.Status.Cleaned = false
+			result.Dirty = true
+			return
+		}
+		if ironicNode.Maintenance {
+			p.log.Info("clearing maintenance flag")
+			return p.setMaintenanceFlag(ironicNode, false)
+		}
+		return p.changeNodeProvisionState(
+			ironicNode,
+			nodes.ProvisionStateOpts{Target: nodes.TargetManage},
+		)
+
+	default:
+		// wait states like cleaning and clean wait
+		p.log.Info("waiting for host to become manageable",
+			"state", ironicNode.ProvisionState,
+			"deploy step", ironicNode.DeployStep)
+		// Avoid start manual cleaning success but set Cleaned flag failed
+		if !p.host.Status.Cleaned {
+			p.host.Status.Cleaned = true
+		}
```

Comment:
What's the reason for this?

Reply:
Because if **start manual clean** succeeds but update **cr** fails, when **manual clean** finish we will start manual clean again, by this way can avoid a part of the above situations.